### PR TITLE
Python 3/Plone 5.2

### PR DESCRIPTION
**Pull request by reinhardt**
I'll throw this out there for discussion - this is not (yet) fit to merge but it gets the test failures down to a handful and seems to work fine in a Plone 5.2 instance (no extensive manual tests yet, though). It contains the work by derFreitag and the plone-5.2-compat and collective-indexing-wip branches.

It still needs cross-checking in Python 2 and older Plone versions. That may also lead to a decision about whether to support all versions in one code base or to branch out, see #210.

I couldn't yet make sense of the failures of testReindexPathIndex and test_retry_on_conflict. Maybe we can have a chat about those at some point (@tisto?).

### CODE REVIEW

**Comment by reinhardt** [comment]
Travis is having trouble because it's not using the python 3 compatible version of plone.restapi. Locally I do - will try to port it to travis.

**Comment by reinhardt** [comment]
OK, that's better. Of the 6 failures/errors, two are probably caused by the same thing (the status 500 ones) and one seems easy to fix but needs a decision how to do it (showinsearch).

I haven't looked at the robot tests yet.

**Comment by tisto** [comment]
@reinhardt let's do a call next week, shall we? I'd be interested to push this further!

**Comment by pbauer** [comment]
I found out that this branch no longer runs Python 2. Is that intentional? Would fixes for Python 2 be welcome?

File: base.cfg [hunk 1]
```diff
@@ -15,10 +15,10 @@ parts +=
     check-readme
 develop = .
 
-# auto-checkout = kitconcept.recipe.solr
+auto-checkout = plone.restapi
 
 [sources]
-plone.restapi = git https://github.com/plone/plone.restapi.git
+plone.restapi = git https://github.com/plone/plone.restapi.git branch=python3_updated
```
Comment (tisto) [inline base.cfg]: @reinhardt could you elaborate why we use the python3_updated branch here? That branch has been merged into plone.restapi long time ago and all recent releases should be python3 compatible.

**Comment by reinhardt** [comment]
@tisto, a call sounds good, yes!

@pbauer, I've been concentrating on the python 3 side and have taken a few short cuts. I'm working on restoring python 2 support. Help is appreciated of course! We should coordinate a bit to avoid conflicts.

File: base.cfg (hunk 1 above)
Comment (reinhardt) [reply base.cfg]: I originally thought plone.restapi master was not python 3 compatible, but at the second glance it looks more like collective.solr imports need some adaptation to work with the latest plone.restapi code.

Comment (tisto) [reply base.cfg]: @reinhardt yeah. We refactored some internals in plone.restapi. Just pin plone.restapi to 3.9.0: https://github.com/collective/collective.solr/commit/50f49e50997620f72f8ea6daea78a8a57392499b

We can look into the plone.restapi upgrade later...

**Comment by tisto** [comment]
@reinhardt could you send me your email address to tisto@plone.org? I will be on holidays next week. Maybe we can find a time in the next days to do a call.

**Comment by reinhardt** [comment]
OK, a02e2bb gets all unit tests green for me locally. Let's hope travis agrees.

For the robot tests I'll have to look into setting up chromedriver, which I might postpone for a bit.

**Comment by reinhardt** [comment]
Well, that wasn't too hard. All green now!

**Comment by tisto** [comment]
@reinhardt w00t!!!

File: src/collective/solr/profiles/default/registry.xml [hunk 2]
```diff
@@ -11,7 +11,7 @@
     <value key="max_results">10000000</value>
     <value key="exclude_user">False
     </value>
-    <value key="search_pattern">+(Title:{value}^5 OR Description:{value}^2 OR SearchableText:{value} OR SearchableText:({base_value}) OR searchwords:({base_value})^1000) +showinsearch:True
+    <value key="search_pattern">+(Title:{value}^5 OR Description:{value}^2 OR SearchableText:{value} OR SearchableText:({base_value}) OR searchwords:({base_value})^1000) -showinsearch:False
```
Comment (tisto) [inline src/collective/solr/profiles/default/registry.xml]: @reinhardt could you elaborate why you changed "showinsearch" here?

**Comment by tisto** [comment]
I forgot to mention that I tried Solr/Plone 5.2 locally and everything works like a charm as far as I can tell after a quick review. 

We would use this in two upcoming projects (event though, we don't need to Python 3 / Plone 5.2 support yet). Just to make sure we can help to find issues as early as possible...

File: src/collective/solr/profiles/default/registry.xml (hunk 2 above)
Comment (reinhardt) [reply src/collective/solr/profiles/default/registry.xml]: A new/custom content type may not have the ISolrFields behavior activated. That means the value for "showinsearch" will be missing. If we require it to be True, then this content type will not be searchable with solr. That's probably unexpected, and also a change from before these fields were converted to a behavior. The original way was that showinsearch was True by default, so all instances of new content types would show up in the search unless you explicitly toggled showinsearch. Searching for content that has showinsearch not set to False, i.e. explicitly True or missing altogether, comes pretty close to that I think.

**Comment by reinhardt** [comment]
1. Squashing would be nice I guess, yes. I think I can invest a little more time and do that.
2. I'd like to double check, but the tests do pass in both the Plone 4 and Plone 5 setups. Maybe c.indexing gets pulled in some other way?!
3. Can't think of any. :-)

**Comment by reinhardt** [comment]
Glad I checked - the travis config is botched. The Plone 4 job is actually installing Plone 5. I previously replaced the Plone version in buildout.cfg, unaware that the travis script was looking for this specific version. I'll make the substitution more robust.

**Comment by reinhardt** [comment]
The different Plone versions are now properly tested. I've fixed the missing collective.indexing on travis and also an ImportError in the tests for Plone < 5.2.

Plone 4.3 has the problem that it doesn't support the ISolrFields behavior. We could restore the old Archetypes schemaextender and make both registrations conditional. However, the GenericSetup step that activates the behavior for the default types breaks on Plone 4 and I'm not sure how best to make that one conditional. We could copy the default profile and conditionally register the correct one depending on the Plone version. That's of course a pain to maintain but if we aim to drop support for Plone 4 soon(-ish) then we could maybe put up with that for a while. What do you think?

**Comment by pbauer** [comment]
This branch is still broken with the version of `plone.restapi` that ships with Plone 5.2.

**Comment by reinhardt** [comment]
Thanks for the hint, I forgot that it's still pinned in our own versions.cfg...

**Comment by tisto** [comment]
@reinhardt would be nice if we could keep Plone 4 compatibility. Then I could test the new release with another project of ours. Otherwise, I'd have to create a branch and stick with the old one, which is also fine with me. I leave that decision up to you...

**Comment by reinhardt** [comment]
Is it OK to just remove the SolrFlareSerializer? The tests pass without it, and its base, the BrainSerializer, has been removed as well. Both should be automatically replaced by the summary serializer. But I'm not familiar with plone.restapi and would be grateful if someone could double check.

If that's fine then only the rebase/squash remains to do, and maybe another pair of eyes on this branch.

**Comment by reinhardt** [comment]
OK, now the travis builds are *really* green. :-D

**Comment by tisto** [comment]
@reinhardt Carsten could have a second look if nobody else steps up. He will be back tomorrow. I would squash and merge then and do a 8.0.0a1 release right away (just in case we might need the 7.x branch, which I hope we don't).

Regarding the SolrFlareSerializer, I'd like to hear @sneridagh's opinion since he added that. I don't recall our use case but we need to double check if c.solr works with p.restapi and Volto. Here is the original PR:

https://github.com/plone/plone.restapi/pull/767/files

File: src/collective/solr/profiles/default/types.xml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0"?>
```
Comment (csenger) [inline src/collective/solr/profiles/default/types.xml]: @reinhardt collective.solr does not require plone.app.contenttypes. We still have the Archetypes extender stuff in the package. Does that work with a non p.a.contenttype site and not break the type definitions?

**Comment by tisto** [comment]
@reinhardt I merged this PR because we would like to test the branch in our non-py3 projects and we need another PR for that: https://github.com/collective/collective.solr/pull/220.

We might still have to solve the p.a.contenttypes issue but after that we can cut an alpha release.

**Pull request merged**